pkg/es/config/config.go

```diff
@@ -317,6 +317,11 @@ func GetHTTPRoundTripper(c *Configuration, logger *zap.Logger) (http.RoundTrippe
 		if err != nil {
 			return nil, err
 		}
+		// ignore closer as we cannot pass it to the round tripper or bubble up
+		_, err = c.TLS.ReloadCertificates(ctlsConfig, logger)
```

Comment:
I don't think this should be done in GetHTTPRoundTripper(), this is why you're not able to use the closer correctly.

Reply:
You should probably remove all places where `Certificates` is used and implement a [`tls.Config#GetCertificate`](https://golang.org/pkg/crypto/tls/#Config) / `#GetClientCertificate` instead. This func would store the current contents of the file, updated whenever the file is updated.

```go
    // GetCertificate returns a Certificate based on the given
    // ClientHelloInfo. It will only be called if the client supplies SNI
    // information or if Certificates is empty.
    //
    // If GetCertificate is nil or returns nil, then the certificate is
    // retrieved from NameToCertificate. If NameToCertificate is nil, the
    // best element of Certificates will be used.
    GetCertificate func(*ClientHelloInfo) (*Certificate, error) // Go 1.4
```

Reply:
We can do that. Initially, I didn't want to change the current behavior for the existing consumers of this package.

The bubbling up the closer will be more complicated, it will require changes in the packages using TLS e.g. ES reader will have to accept the closer in the factory method etc.

Reply:
> I don't think this should be done in GetHTTPRoundTripper(), this is why you're not able to use the closer correctly.

@yurishkuro the watcher will be encapsulated with our TLS config struct, hence the closer will be created when the go TLS is created and this happens in the round tripper unless the TLS creation is a two-step process.

Reply:
Yes, I am suggesting to make it two-step process. All storage implementations that use TLS have some sort of start-up logic, e.g. where they establish connections to the backends. That start-up logic has the corresponding Close() path. We can start cert watching there, and close in the Close path.